### PR TITLE
Docker image for Levanter

### PR DESCRIPTION
**Pull request by DwarKapex**
Incorporation of Levanter project (https://github.com/stanford-crfm/levanter) into JAX building system

### CODE REVIEW

**Comment by yhtang** [comment]
Is this PR ready for review? It doesn't have any description.

File: .github/container/levanter-cache-warn.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+YELLOW='\033[0;33m'
+NOCOLOR='\033[0m'
+
+echo -e "${YELLOW}
+WARNING: To speed up the model training it is recommended to cache the datasets
```
Comment (yhtang) [inline .github/container/levanter-cache-warn.sh:7]: The message should ask the user to make sure that `--cache_dir` is a **shared** filesystem across nodes.
In a HPC cluster environment, a common practice is to mount a workspace/project folder from a distributed file system into the designated container path.

File: .github/container/manifest.yaml [hunk 2]
```diff
@@ -117,3 +117,13 @@ maxtext:
   tracking_ref: main
   latest_verified_commit: ec9cdce1928fb36f93e498c8f74f8f130bad3dc2
   mode: git-clone
+levanter:
+  url: https://github.com/stanford-crfm/levanter.git
+  tracking_ref: slurm_cluster_fix
```
Comment (yhtang) [inline .github/container/manifest.yaml:122]: Is the branch already merged into `main`?

File: .github/workflows/_test_levanter.yaml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          docker pull ${{ inputs.JAX_IMAGE }}
+
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.10
+          
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
```
Comment (yhtang) [inline .github/workflows/_test_levanter.yaml:69]: Is this happening on the runner host?

File: .github/workflows/_test_levanter.yaml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+jobs:
+  runner:
+    uses: ./.github/workflows/_runner_ondemand_slurm.yaml
+    with:
+      NAME: "A100-${{ github.run_id }}"
+      LABELS: "A100:${{ github.run_id }}"
+      TIME: "01:00:00"
+    secrets: inherit
+
+  jax-unit-test:
```
Comment (yhtang) [inline .github/workflows/_test_levanter.yaml:31]: The job name still suggests JAX.

File: .github/workflows/_test_levanter.yaml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+          python-version: 3.10
+          
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install flake8 pytest
+
+      - name: Test with pytest
+        run: |
+          XLA_FLAGS=--xla_force_host_platform_device_count=8 PYTHONPATH=tests:src:. pytest tests | tee test-levanter.log
```
Comment (yhtang) [inline .github/workflows/_test_levanter.yaml:74]: Is this running outside a container?

**Review by yhtang — CHANGES_REQUESTED** [review]

File: .github/container/levanter-cache-warn.sh (hunk 1 above)
Comment (DwarKapex) [reply .github/container/levanter-cache-warn.sh:7]: Looks like done

File: .github/workflows/_test_levanter.yaml (hunk 3 above)
Comment (DwarKapex) [reply .github/workflows/_test_levanter.yaml]: Nope, should be inside container

File: .github/workflows/_test_levanter.yaml (hunk 5 above)
Comment (DwarKapex) [reply .github/workflows/_test_levanter.yaml]: Nope, inside

File: .github/container/levanter-cache-warn.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+YELLOW='\033[0;33m'
+NOCOLOR='\033[0m'
+
+echo -e "${YELLOW}
+WARNING: To speed up the model training it is recommended to cache the datasets with:
```
Comment (dlwh) [inline .github/container/levanter-cache-warn.sh:7]: so it's not so much about speeding up (in a "real" run it shouldn't really impact throughput much at all aside from initial warmup). It does make a material difference in unit tests.

 I'll make a more detailed help section and maybe I can PR it to this section?

File: .github/container/levanter-cache-warn.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+When run on distributed system, please make sure that the cache folder <path/to/shared/cache>
+is on a network/shared filesystem across nodes. 
+
+To supply the cache folder to Levanter, use `--data.cache_dir` flag:
+
+    $ docker run -v <path/to/shared/cache>:<path/to/in-docker/cache> <levanter-docker-image> \\
+        python -m levanter.main.train_lm <...> --data.cache_dir <path/to/in-docker/cache>
+
+Alternatevely, you can skip `--data.cache_dir` flag, then the cache folder is `/opt/levanter/cache`
+be default:
```
Comment (dlwh) [inline .github/container/levanter-cache-warn.sh:20]:
```suggestion
by default:
```

File: .github/container/levanter-cache-warn.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+When run on distributed system, please make sure that the cache folder <path/to/shared/cache>
+is on a network/shared filesystem across nodes. 
+
+To supply the cache folder to Levanter, use `--data.cache_dir` flag:
+
+    $ docker run -v <path/to/shared/cache>:<path/to/in-docker/cache> <levanter-docker-image> \\
+        python -m levanter.main.train_lm <...> --data.cache_dir <path/to/in-docker/cache>
+
+Alternatevely, you can skip `--data.cache_dir` flag, then the cache folder is `/opt/levanter/cache`
```
Comment (dlwh) [inline .github/container/levanter-cache-warn.sh:19]:
```suggestion
Alternatively, you can skip `--data.cache_dir` flag, then the cache folder is `/opt/levanter/cache`
```

File: .github/container/levanter-cache-warn.sh [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    $ python -m levanter.main.cache_dataset --id <dataset_id> --cache_dir <path/to/shared/cache>
+
+When run on distributed system, please make sure that the cache folder <path/to/shared/cache>
+is on a network/shared filesystem across nodes. 
+
+To supply the cache folder to Levanter, use `--data.cache_dir` flag:
+
+    $ docker run -v <path/to/shared/cache>:<path/to/in-docker/cache> <levanter-docker-image> \\
+        python -m levanter.main.train_lm <...> --data.cache_dir <path/to/in-docker/cache>
+
```
Comment (dlwh) [inline .github/container/levanter-cache-warn.sh:18]:
```suggestion

This directory can be any network-readable path, a path in an object store (e.g. GCS). or a path to a pre-built cache.

```

**Review by dlwh — COMMENTED** [review]
seems good to me, though I didn't attempt to fully decode all the docker-y yaml-y stuff. I really appreciate it!

I'll add some better docs for the cache stuff and PR a link to that.

File: .github/container/levanter-cache-warn.sh (hunk 6 above)
Comment (yhtang) [reply .github/container/levanter-cache-warn.sh]: @dlwh Sounds like a plan

File: .github/container/levanter-cache-warn.sh (hunk 9 above)
Comment (DwarKapex) [reply .github/container/levanter-cache-warn.sh:18]: Added

File: .github/container/levanter-cache-warn.sh (hunk 8 above)
Comment (DwarKapex) [reply .github/container/levanter-cache-warn.sh:19]: Fixed

File: .github/container/levanter-cache-warn.sh (hunk 7 above)
Comment (DwarKapex) [reply .github/container/levanter-cache-warn.sh:20]: Done

File: .github/container/levanter-cache-warn.sh (hunk 6 above)
Comment (DwarKapex) [reply .github/container/levanter-cache-warn.sh:7]: > I'll make a more detailed help section and maybe I can PR it to this section?

No objections from my side. 
@yhtang What do you think about it?

**Review by yhtang — APPROVED** [review]
LGTM. We'll merge once the tests pass.

**Review by yhtang — APPROVED** [review]

**Pull request merged**